[InstCombine] Introduce `foldICmpBinOpWithConstantViaTruthTable` folding

antoniofrighetto opened the pull request:
Match icmps of binops where both operands are select with constant arms, i.e., `icmp pred (select A ? C1 : C2) binop (select B ? C3 : C4), C5`. Fold such patterns by creating a truth table of the possible four constant variants, and materialize back the optimal logic from it via `createLogicFromTable` helper. This also generalizes an existing fold, which has therefore been dropped.

Proofs: https://alive2.llvm.org/ce/z/NS7Vzu.

Fixes: https://github.com/llvm/llvm-project/issues/138212.

nikic reviewed (commented):
Does this subsume the fold at https://github.com/llvm/llvm-project/blob/948bffa951cf143da5f4caa461b25ce76fa137d0/llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp#L1969-L1991?

nikic commented on llvm/test/Transforms/InstCombine/icmp-binop.ll:
```diff
@@ -359,3 +359,99 @@ define i1 @test_icmp_sgt_and_negpow2_invalid_c(i32 %add) {
+; CHECK-NEXT:    ret <2 x i1> [[CMP]]
+;
+  %s1 = select <2 x i1> %a, <2 x i64> <i64 65536, i64 65536>, <2 x i64> zeroinitializer
+  %s2 = select <2 x i1> %b, <2 x i64> <i64 256, i64 256>, <2 x i64> zeroinitializer
```
Can you please add a test where the select condition is a constant but the values are vectors?

antoniofrighetto replied:
If the condition is constant the select will be constant folded, if arms are values, fold won't be carried out?

nikic replied:
Oops, I mean to say "where the select condition is a scalar but the values are vectors".

antoniofrighetto replied:
Added, thanks, updated by bailing out (can't splat here as invalid: https://alive2.llvm.org/ce/z/WVqkCC).

nikic commented on llvm/test/Transforms/InstCombine/icmp-binop.ll:
```diff
@@ -359,3 +359,99 @@ define i1 @test_icmp_sgt_and_negpow2_invalid_c(i32 %add) {
+  %s1 = select i1 %a, i64 0, i64 65536
+  %s2 = select i1 %b, i64 0, i64 256
+  %or = or i64 %s1, %s2
+  call void @use64(i64 %or)
```
Can you also add a multi-use test that *does* fold because it only generates one instruction?

antoniofrighetto commented:
> Does this subsume the fold at
> 
> https://github.com/llvm/llvm-project/blob/948bffa951cf143da5f4caa461b25ce76fa137d0/llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp#L1969-L1991
> 
> ?

Right, thanks for catching it, PR description updated.

nikic commented on llvm/test/Transforms/InstCombine/icmp-binop.ll:
```diff
@@ -1,5 +1,5 @@
 ; NOTE: Assertions have been autogenerated by utils/update_test_checks.py
-; RUN: opt < %s -passes=instcombine -S | FileCheck %s
+; RUN: opt < %s -passes="instcombine<no-verify-fixpoint>" -S | FileCheck %s
```
Why is this needed?

antoniofrighetto replied:
Got `LLVM ERROR: Instruction Combining on icmp_eq_or_of_selects_with_scalar_cond_constant_vectorized did not reach a fixpoint after 1 iterations.` after running the test. Did I miss something in the code or is this an orthogonal issue?

dtcxzyw commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -3110,6 +3086,45 @@ static Value *createLogicFromTable(const std::bitset<4> &Table, Value *Op0,
+    // Synthesize optimal logic.
+    if (auto *Cond =
+            createLogicFromTable(Table, A, B, Builder, BO->hasOneUse());
+        Cond && Cmp.getType() == Cond->getType())
```
If `Cmp.getType() != Cond->getType()`, the newly created instruction will not be used. It is the root cause of https://github.com/llvm/llvm-project/pull/139109/files#r2083295338.

dtcxzyw reviewed (approved):
LGTM.

nikic commented on llvm/lib/Transforms/InstCombine/InstCombineCompares.cpp:
```diff
@@ -3110,6 +3086,45 @@ static Value *createLogicFromTable(const std::bitset<4> &Table, Value *Op0,
+            m_Select(m_Value(A), m_Constant(C1), m_Constant(C2))) &&
+      match(BO->getOperand(1),
+            m_Select(m_Value(B), m_Constant(C3), m_Constant(C4))) &&
+      Cmp.getType() == A->getType()) {
```
Early return?

nikic reviewed (approved):
LGTM

antoniofrighetto committed:
[InstCombine] Precommit tests for PR139109 (NFC)

antoniofrighetto committed:
[InstCombine] Introduce `foldICmpBinOpWithConstantViaTruthTable` folding

Match icmps of binops where both operands are select with constant arms,
i.e., `icmp pred (select A ? C1 : C2) binop (select B ? C3 : C4), C5`.
Fold such patterns by creating a truth table of the possible four
constant variants, and materialize back the optimal logic from it via
`createLogicFromTable` helper. This also generalizes an existing fold,
which has therefore been dropped.

Proofs: https://alive2.llvm.org/ce/z/NS7Vzu.

Fixes: https://github.com/llvm/llvm-project/issues/138212.
